### PR TITLE
Introduce `Fixed128` from ORML to `sp_arithmetic`

### PR DESCRIPTION
**Pull request by xlc**
Supersede #5606

Initially developed at https://github.com/laminar-protocol/open-runtime-module-library/blob/master/utilities/src/fixed_128.rs

Only checked and saturating ops are implemented and no panic ops.

TODO:

- [ ] test for `saturating_pow`

### CODE REVIEW

**Comment by gavofyork** [comment]
build failing?

**Comment by shawntabrizi** [comment]
Please also update `primitives/runtime/src/lib.rs` to re-export Fixed128.

I would also like to take a stab and update Fixed64 to give it the same API as in your Fixed128 implementation... maybe a shared trait `FixedPointNumber`? Can I make those changes?

**Comment by xlc** [comment]
I think we can have something similar to `PerThing` and `PerX` and use macro to generate the implementations. So we have a `FixedPointNumber` and some macro to generate Fixed32/64/128. But let's start with this one and refactor it later.

File: primitives/arithmetic/src/fixed128.rs [hunk 1]
```diff
@@ -0,0 +1,675 @@
...
+	pub fn is_positive(&self) -> bool {
+		self.0.is_positive()
+	}
+
+	pub fn is_negative(&self) -> bool {
+		self.0.is_negative()
+	}
+}
+
+impl Saturating for Fixed128 {
```
Comment (shawntabrizi) [inline primitives/arithmetic/src/fixed128.rs:219]:
```suggestion
/// Note that this is a standard, _potentially-panicking_, implementation. Use `Saturating` trait
/// for safe addition.
impl ops::Add for Fixed128 {
	type Output = Self;

	fn add(self, rhs: Self) -> Self::Output {
		Self(self.0 + rhs.0)
	}
}

/// Note that this is a standard, _potentially-panicking_, implementation. Use `Saturating` trait
/// for safe subtraction.
impl ops::Sub for Fixed128 {
	type Output = Self;

	fn sub(self, rhs: Self) -> Self::Output {
		Self(self.0 - rhs.0)
	}
}

impl Saturating for Fixed128 {
```

Any reason not to include basic Add/Sub traits? Makes it usable in many scenarios, especially tests.

**Comment by shawntabrizi** [comment]
TODO follow up here is to clean up the API between Fixed64 and Fixed128, for example we have `Fixed64::into_inner()` where we have now `Fixed128::deconstruct()`.

File: primitives/arithmetic/src/fixed128.rs (hunk 1 above)
Comment (xlc) [reply primitives/arithmetic/src/fixed128.rs:219]: we just don't want to introduce any new possibilities of panics.

for tests we are currently use `value.checked_add(other).unwrap()`

Comment (shawntabrizi) [reply primitives/arithmetic/src/fixed128.rs:219]: Practically speaking though, not implementing these traits can be a major pain in the ass for regular usage. My use case specifically was using `assert_eq_error_rate` which needed the type to support `Add`. I mean there are other ways to fix this problem, but I dont think you are really causing any more or less problems implementing these very standard traits.

Comment (shawntabrizi) [reply primitives/arithmetic/src/fixed128.rs:219]: Maybe add it with the `test` cfg

File: primitives/arithmetic/src/fixed128.rs [hunk 2]
```diff
@@ -0,0 +1,675 @@
...
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+use codec::{Decode, Encode};
+use primitive_types::U256;
```
Comment (kianenigma) [inline primitives/arithmetic/src/fixed128.rs:18]: Maybe we should investigate if we can avoid this or not. We have so far avoided including these primitive types in the default substrate runtime. We have our own minimal big-int primitives which should be enough for this case.

Comment (xlc) [reply primitives/arithmetic/src/fixed128.rs:18]: how about feature flag it for now and refactor it later?

Comment (dvdplm) [reply primitives/arithmetic/src/fixed128.rs:18]: Is switching to using the [`uint` crate](https://crates.io/crates/uint) and use `construct_uint!` directly? Seems like `U256` is used here just to handle overflows and the type is not exported in the public interface.

**Pull request merged**